src/cipher.jl

```diff
@@ -33,120 +18,146 @@ immutable NettleCipher
-    const_ctx_size = nh.context_size
-    const_block_size = nh.block_size
-    const_key_size = nh.key_size
-
-    # Save the function pointers as well
-    fptr_set_encrypt_key = nh.set_encrypt_key
-    fptr_set_decrypt_key = nh.set_decrypt_key
-    fptr_encrypt = nh.encrypt
-    fptr_decrypt = nh.decrypt
-
-
-    # First, create the type itself
-    @eval immutable $name <: CipherAlgorithm; end
-
-    # Next, record all the important information about this cipher algorithm
-    @eval ctx_size(::Type{$name}) = $(convert(Int,const_ctx_size))
-    @eval cipher_type(::Type{$name}) = $nhptr
-    @eval block_size(::Type{$name}) = $(convert(Int,const_block_size))
-    @eval key_size(::Type{$name}) = $(convert(Int,const_key_size))
-
-    # Generate the constructors and encrypt/decrypt functions while we're at it!
-    # Since we have the function pointers from nh, we'll use those
-    @eval function CipherEncrypt(::Type{$name},key::Union(AbstractString,Vector{UInt8}))
-        length(key) != key_size($name) && error("Key must be $(key_size($name)) bytes long")
-        ctx = Array(UInt8, ctx_size($name))
-        if nettle_major_version >= 3
-            ccall($fptr_set_encrypt_key, Void, (Ptr{Void}, Ptr{UInt8}),
-                  ctx, pointer(key))
-        else
-            ccall($fptr_set_encrypt_key, Void, (Ptr{Void}, Cuint, Ptr{UInt8}),
-                  ctx, length(key), pointer(key))
+function get_cipher_types()
+    # If we have already gotten the hash types from libnettle, don't query again
+    if isempty(_cipher_types)
+        cipher_idx = 1
+        # nettle_ciphers is an array of pointers ended by a NULL pointer, continue reading hash types until we hit it
+        while( true )
+            ncptr = unsafe_load(cglobal(("nettle_ciphers",nettle),Ptr{Ptr{Void}}),cipher_idx)
+            if ncptr == C_NULL
+                break
+            end
+            cipher_idx += 1
+            nc = unsafe_load(convert(Ptr{NettleCipher}, ncptr))
+            cipher_type = NettleCipherType(nc)
+            _cipher_types[cipher_type.name] = cipher_type
         end
-        CipherEncrypt{$name}(ctx)
     end
+    return _cipher_types
+end
 
-    @eval function CipherDecrypt(::Type{$name},key::Union(AbstractString,Vector{UInt8}))
-        length(key) != key_size($name) && error("Key must be $(key_size($name)) bytes long")
-        ctx = Array(UInt8, ctx_size($name))
-        if nettle_major_version >= 3
-            ccall($fptr_set_decrypt_key, Void, (Ptr{Void}, Ptr{UInt8}),
-                  ctx, pointer(key))
-        else
-            ccall($fptr_set_decrypt_key, Void, (Ptr{Void}, Cuint, Ptr{UInt8}),
-                  ctx, length(key), pointer(key))
-        end
-        CipherDecrypt{$name}(ctx)
-    end
 
-    @eval function decrypt(state::CipherDecrypt{$name},source)
-        result = Array(UInt8,length(source))
-        decrypt!(state,result,source)
-        result
+function EncryptionAlgorithm(name::AbstractString, key)
```

Comment:
This is capitalized like a type constructor, but it doesn't return a type of the same name.  Wouldn't it be more Julian to actually make this a constructor?

It could be less verbose by calling the function and the type `Encrypter` and `Decrypter`.